waiter/src/waiter/descriptor.clj

```diff
@@ -0,0 +1,268 @@
+          (-> (select-keys descriptor [:passthrough-headers :waiter-headers])
+              (assoc :sources new-sources)
+              (sd/merge-service-description-and-id
+                kv-store service-id-prefix username metric-group-mappings service-description-builder service-approved?)
+              (sd/merge-suspended kv-store)))))))
+
+(let [request->descriptor-timer (metrics/waiter-timer "core" "request->descriptor")]
+  (defn request->descriptor
+    "Extract the service descriptor from a request.
+     It also performs the necessary authorization."
+    [assoc-run-as-user-approved? can-run-as? start-new-service-fn fallback-state-atom kv-store metric-group-mappings
+     search-history-length service-description-builder service-description-defaults service-id-prefix token-defaults
+     waiter-hostnames {:keys [request-time] :as request}]
+    (timers/start-stop-time!
+      request->descriptor-timer
+      (let [auth-user (:authorization/user request)
+            service-approved? (fn service-approved? [service-id] (assoc-run-as-user-approved? request service-id))
+            latest-descriptor (compute-descriptor
+                                service-description-defaults token-defaults service-id-prefix kv-store waiter-hostnames
+                                request metric-group-mappings service-description-builder service-approved?)
+            latest-service-id (:service-id latest-descriptor)
+            descriptor->previous-descriptor
+            (fn descriptor->previous-descriptor-fn
+              [descriptor]
+              (descriptor->previous-descriptor
+                kv-store service-id-prefix token-defaults metric-group-mappings service-description-builder
+                service-approved? auth-user descriptor))
+            fallback-state @fallback-state-atom
+            descriptor (resolve-descriptor
+                         descriptor->previous-descriptor search-history-length request-time fallback-state latest-descriptor)
+            {:keys [service-authentication-disabled service-description service-preauthorized]} descriptor
+            {:strs [run-as-user permitted-user]} service-description]
+        (when-not (or service-authentication-disabled
+                      service-preauthorized
+                      (and auth-user (can-run-as? auth-user run-as-user)))
+          (throw (ex-info "Authenticated user cannot run service"
+                          {:authenticated-user auth-user
+                           :run-as-user run-as-user
+                           :status 403})))
+        (when-not (request-authorized? auth-user permitted-user)
+          (throw (ex-info "This user isn't allowed to invoke this service"
+                          {:authenticated-user auth-user
+                           :service-description service-description
+                           :status 403})))
+        (let [fallback-service-id (:service-id descriptor)]
```

Comment:
Does this logic make more sense inside `wrap-descriptor` instead of `request->descriptor`?

Reply:
Done.